pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -20,108 +20,194 @@ import (
+	return &http.Server{
+		Addr:    addr,
+		Handler: handler,
 	}
-	sendHealthCheckResponse(rw, http.StatusServiceUnavailable, "0 local Endpoints are alive")
 }
 
-// handleMutationRequest - receive requests to mutate the table entry for a service
-func (h *proxyHC) handleMutationRequest(req *proxyMutationRequest) {
-	numEndpoints := len(*req.endpointUids)
-	glog.V(4).Infof("LB service health check mutation request Service: %s - %d Endpoints %v",
-		req.serviceName, numEndpoints, (*req.endpointUids).List())
-	if numEndpoints == 0 {
-		if _, ok := h.serviceEndpointsMap.Get(req.serviceName.String()); ok {
-			glog.V(4).Infof("Deleting endpoints map for service %s, all local endpoints gone", req.serviceName.String())
-			h.serviceEndpointsMap.Delete(req.serviceName.String())
+var _ HTTPServerFactory = stdHTTPServerFactory{}
+
+type server struct {
+	listener    Listener
+	httpFactory HTTPServerFactory
+
+	lock      sync.Mutex
+	services  map[types.NamespacedName]*service
+	endpoints map[types.NamespacedName]int // number of local endpoints for a service
+}
+
+func (hcs *server) SyncServices(newServices map[types.NamespacedName]uint16) error {
+	hcs.lock.Lock()
+	defer hcs.lock.Unlock()
+
+	// Remove any that are not needed any more.
+	for nsn, svc := range hcs.services {
+		if port, found := newServices[nsn]; !found || port != svc.port {
+			glog.V(2).Infof("closing healthcheck %q on port %d", nsn.String(), svc.port)
+			svc.listener.Close()
+			delete(hcs.services, nsn)
+			delete(hcs.endpoints, nsn)
 		}
-		return
 	}
-	var entry *serviceEndpointsList
-	e, exists := h.serviceEndpointsMap.Get(req.serviceName.String())
-	if exists {
-		entry = e.(*serviceEndpointsList)
-		if entry.endpoints.Equal(*req.endpointUids) {
-			return
-		}
-		// Compute differences just for printing logs about additions and removals
-		deletedEndpoints := entry.endpoints.Difference(*req.endpointUids)
-		newEndpoints := req.endpointUids.Difference(*entry.endpoints)
-		for _, e := range newEndpoints.List() {
-			glog.V(4).Infof("Adding local endpoint %s to LB health check for service %s",
-				e, req.serviceName.String())
+
+	// Add any that are needed.
+	for nsn, port := range newServices {
+		if hcs.services[nsn] != nil {
+			glog.V(3).Infof("existing healthcheck %q on port %d", nsn.String(), port)
+			continue
 		}
-		for _, d := range deletedEndpoints.List() {
-			glog.V(4).Infof("Deleted endpoint %s from service %s LB health check (%d endpoints left)",
-				d, req.serviceName.String(), len(*entry.endpoints))
+
+		glog.V(2).Infof("opening healthcheck %q on port %d", nsn.String(), port)
+		svc := &service{port: port}
+		addr := fmt.Sprintf(":%d", port)
+		svc.server = hcs.httpFactory.New(addr, hcHandler{name: nsn, hc: hcs})
+		var err error
+		svc.listener, err = hcs.listener.Listen(addr)
+		if err != nil {
+			glog.Errorf("failed to start healthcheck %q on port %d: %v", nsn.String(), port, err)
```

Comment:
Is there any good way to exposed this? Like a event? (But no corresponding event for success though) No listener means the endpoints on the node could not serving traffic.